spec/models/contribution_spec.rb

```diff
@@ -251,4 +251,77 @@
       it{ is_expected.to eq(0) }
     end
   end
+
+  describe '#change_state_by_transaction' do
+    before do
+      CatarseSettings[:email_payments] = 'financial@administrator.com'
+
+      create(:user, email: CatarseSettings[:email_payments])
+    end
+
+    context 'when transaction status is waiting payment' do
+      it 'contribution state should change to waiting confirmation' do
+        contribution = create(:contribution, state: 'pending')
+
+        expect { contribution.change_state_by_transaction('waiting_payment') }
+          .to change { contribution.reload.state }
+          .from('pending').to('waiting_confirmation')
+      end
+    end
+
+    context 'when transaction status is paid' do
+      it 'contribution state should change to confirmed' do
+        contribution = create(:contribution, state: 'pending')
+
+        expect { contribution.change_state_by_transaction('paid') }
+          .to change { contribution.reload.state }
+          .from('pending').to('confirmed')
+      end
+    end
+
+    context 'when transaction status is authorized' do
+      it 'contribution state should change to confirmed' do
+        contribution = create(:contribution, state: 'pending')
+
+        expect { contribution.change_state_by_transaction('authorized') }
+          .to change { contribution.reload.state }
+          .from('pending').to('confirmed')
+      end
+    end
+
+    context 'when transaction status is pending refund' do
+      it 'contribution state should change to requested refund' do
+        contribution = create(:contribution, state: 'confirmed')
+        
+        allow(contribution.user).to receive(:credits)
```

Comment:
até onde sei, n falha se n for float.
Então... é "testado" aqui: https://github.com/juntos-com-vc/juntos.com.vc/blob/92afbf43670ddebd4bab26c30346934b29a8fd35/spec/models/user_spec.rb#L371
essas aspas estão aí por motivo de `xit` no teste :/